x-pack/agent/pkg/reporter/fleet/reporter.go

```diff
@@ -41,136 +40,111 @@ func (e *event) Message() string {
-	) (*http.Response, error)
-}
-
 // Reporter is a reporter without any effects, serves just as a showcase for further implementations.
 type Reporter struct {
-	logger         *logger.Logger
-	queue          []reporter.Event
-	qlock          sync.Mutex
-	ticker         *time.Ticker
-	threshold      int
-	droppedCounter int
-	checkingCmd    checkinExecutor
-	closeChan      chan struct{}
-	closeOnce      sync.Once
+	logger    *logger.Logger
+	queue     []fleetapi.SerializableEvent
+	qlock     sync.Mutex
+	threshold int
+	lastAck   time.Time
 }
 
 type agentInfo interface {
 	AgentID() string
 }
 
 // NewReporter creates a new fleet reporter.
-func NewReporter(agentInfo agentInfo, l *logger.Logger, c *ManagementConfig, client remoteClient) (*Reporter, error) {
-	checkinClient := fleetapi.NewCheckinCmd(agentInfo, client)
-
-	frequency := time.Duration(c.ReportingCheckFrequency) * time.Second
+func NewReporter(agentInfo agentInfo, l *logger.Logger, c *ManagementConfig) (*Reporter, error) {
 	r := &Reporter{
-		queue:       make([]reporter.Event, 0),
-		ticker:      time.NewTicker(frequency),
-		logger:      l,
-		checkingCmd: checkinClient,
-		threshold:   c.Threshold,
-		closeChan:   make(chan struct{}),
+		queue:     make([]fleetapi.SerializableEvent, 0),
+		logger:    l,
+		threshold: c.Threshold,
 	}
 
-	go r.reportLoop()
 	return r, nil
 }
 
-// Report in noop reporter does nothing.
+// Report enqueue event into reporter queue.
 func (r *Reporter) Report(e reporter.Event) error {
 	r.qlock.Lock()
 	defer r.qlock.Unlock()
 
-	r.queue = append(r.queue, e)
+	r.queue = append(r.queue, &event{
+		EventType: e.Type(),
+		Ts:        fleetapi.Time(e.Time()),
+		SubType:   e.SubType(),
+		Msg:       e.Message(),
+		Payload:   e.Payload(),
+		Data:      e.Data(),
+	})
+
 	if r.threshold > 0 && len(r.queue) > r.threshold {
+		// drop some low importance event if needed
 		r.dropEvent()
 	}
-	return nil
-}
 
-// Close stops all the background jobs reporter is running.
-// Guards agains panic of closing channel multiple times.
-func (r *Reporter) Close() error {
-	r.closeOnce.Do(func() { close(r.closeChan) })
 	return nil
 }
 
-func (r *Reporter) reportLoop() {
-	for {
-		select {
-		case <-r.ticker.C:
-		case <-r.closeChan:
-			r.logger.Info("stop received, cancelling the fleet report loop")
-			return
-		}
+// Events returns a list of event from a queue and a ack function
+// which clears those events once caller is done with processing.
+func (r *Reporter) Events() ([]fleetapi.SerializableEvent, func()) {
+	r.qlock.Lock()
+	defer r.qlock.Unlock()
 
-		// report all events up to this point
-		r.qlock.Lock()
-		batch := r.queueCopy()
-		r.droppedCounter = 0
-		r.qlock.Unlock()
+	cp := r.queueCopy()
 
-		if err := r.reportBatch(batch); err != nil {
-			r.logger.Errorf("failed to report event batch: %v", err)
-			continue
-		}
+	ackFn := func() {
+		// as time is monotonic and this is on single machine this should be ok.
+		r.clear(cp, time.Now())
+	}
```

Comment:
I think there is a problem here:

1.We take a copy of the events in the queue.
2. We create closure.
3. The time used to clear items out is generated in the closure when the function is called, so possibly events added after in the queue will be still be removed.


It's probably better use to use index position or sequence number on the events?

Reply:
good point with evaluation i will move it outside of closure, index position is tricky as we are dropping events. we do not know what is at that location at the moment of calling ack. with ack calls out of order this can become a problem.

so i went with the clock. if you call ack which was created after ack1, then all items related to ack1 were also included in ack (as ack1 was not called at the time of creation of ack) so ack() will remove both ack and ack1 items. then if you call ack1 and you know items related were already removed because later ack was called before.

on the other hand if you have 2 sets ack1 and ack2 where ack2 includes ack1 then index of ack2 would be idx1+idx2 and after calling ack1 and relying on indexes you would get data loss of idx1 items.

Reply:
i hope it makes sense :D

Reply:
Let's keep it that way, I think we will revisit that shortly see my [comment](https://github.com/elastic/beats/pull/15237/files#r362079722) concerning batching. Let's make incremental steps for implementation.